lib/vector-buffers/src/topology/channel/limited_queue.rs

```diff
@@ -213,462 +187,441 @@ pub struct LimitedReceiver<T> {
+                return None;
             }
 
             // We're not closed, so we need to wait for a writer to tell us they made some
             // progress.  This might end up being a spurious wakeup since `Notify` will
-            // store up to one wakeup that gets consumed by the next call to `poll_notify`,
-            // but alas.
-            ready!(self.inner.read_waker.poll_notify(cx));
+            // store a wake-up if there are no waiters, but oh well.
+            self.inner.read_waker.notified().await;
         }
     }
 
-    pub fn close(&mut self) {
-        self.inner.limiter.close();
-    }
-}
-
-impl<T: Bufferable> Stream for LimitedReceiver<T> {
-    type Item = T;
-
-    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<T>> {
-        Pin::into_inner(self).poll_next(cx)
-    }
-}
-
-impl<T> Clone for LimitedSender<T> {
-    fn clone(&self) -> Self {
-        self.sender_count.fetch_add(1, Ordering::SeqCst);
-
-        Self {
-            inner: self.inner.clone(),
-            sender_count: Arc::clone(&self.sender_count),
-            slot: None,
-        }
+    pub fn into_stream(self) -> Pin<Box<dyn Stream<Item = T> + Send>> {
+        let mut receiver = self;
+        Box::pin(stream! {
+            while let Some(item) = receiver.next().await {
+                yield item;
+            }
+        })
     }
 }
 
-impl<T> Drop for LimitedSender<T> {
+impl<T> Drop for LimitedReceiver<T> {
     fn drop(&mut self) {
-        // If we're the last sender to drop, close the semaphore on our way out the door.
-        if self.sender_count.fetch_sub(1, Ordering::SeqCst) == 1 {
-            self.inner.limiter.close();
-            self.inner.read_waker.as_ref().notify_one();
-        }
+        // Notify senders that the channel is now closed by closing the semaphore.  Any pending
+        // acquisitions will be awoken and notified that the semaphore is closed, and further new
+        // sends will immediately see the semaphore is closed.
+        self.inner.limiter.close();
     }
 }
 
 pub fn limited<T>(limit: usize) -> (LimitedSender<T>, LimitedReceiver<T>) {
     let inner = Inner {
         data: Arc::new(ArrayQueue::new(limit)),
         limit,
-        limiter: PollSemaphore::new(Arc::new(Semaphore::new(limit))),
-        read_waker: PollNotify::new(Arc::new(Notify::new())),
-        write_waker: PollNotify::new(Arc::new(Notify::new())),
+        limiter: Arc::new(Semaphore::new(limit)),
+        read_waker: Arc::new(Notify::new()),
     };
 
     let sender = LimitedSender {
         inner: inner.clone(),
         sender_count: Arc::new(AtomicUsize::new(1)),
-        slot: None,
     };
     let receiver = LimitedReceiver { inner };
 
     (sender, receiver)
 }
 
-#[cfg(test)]
-mod tests {
-    use futures::future::poll_fn;
-    use tokio_test::{assert_pending, assert_ready, task::spawn};
+// TODO: uncomment and fix these, probably using `tokio_test::task::spawn` to drive the actual
```

Comment:
Are these a pre-merge TODO?

Reply:
Absolutely, yes.  One of a few remaining items listed in the PR description.